PDFBOX-5872: Support imageio-openjpeg for JPEG2000 reading

Googulator opened the pull request:
The imageio-openjpeg doesn't implement raster reading, but PDFBox no longer relies on raster support, so just removing that check makes imageio-openjpeg work with PDFBox when the required native library is installed.

However, when the native library is not installed, imageio-openjpeg indicates that it's not usable by returning null as the image reader. This needs to be checked to avoid NPEs merely from this library being installed without its native component.

Googulator committed:
Support imageio-openjpeg for JPEG2000 reading

The imageio-openjpeg doesn't implement raster reading, but PDFBox no longer relies on raster support, so just removing that check makes imageio-openjpeg work with PDFBox when the required native library is installed.

However, when the native library is not installed, imageio-openjpeg indicates that it's not usable by returning null as the image reader. This needs to be checked to avoid NPEs merely from this library being installed without its native component.

Googulator committed:
PDFBOX-5872: request an image reader with raster support only when needed

Googulator committed:
PDFBOX-5872: don't leak rejected image readers

THausherr commented:
Now you have 5 files changed and a method that is harder to read.
Please watch this: https://www.youtube.com/watch?v=2Q9GRPxqCAk

Googulator committed:
PDFBOX-5872: eliminate boolean argument from findImageReader

Instead, it's split into findImageReader and findRasterReader.

JBIG2 also uses non-raster reads only, so requiring canReadRaster()
from the JBIG2 reader is unneessary, although I'm not aware of any
JBIG2 reader that would be excluded by canReadRaster().

Googulator commented:
Making the non-raster version private would mean leaving JBIG2 with an unnecessary restriction to readers with raster support only.

Now, unlike JPEG2000, I'm not aware of any problematic JBIG2 reader, but semantically it's still more correct to accept non-raster-capable readers.
